libyul/optimiser/Suite.cpp

```diff
@@ -429,6 +246,59 @@ map<char, string> const& OptimiserSuite::stepAbbreviationToNameMap()
 	return lookupTable;
 }
 
+void OptimiserSuite::runSequence(string const& _stepAbbreviations, Block& _ast)
+{
+	string input = _stepAbbreviations;
+	boost::remove_erase(input, ' ');
+	boost::remove_erase(input, '\n');
+
+	bool insideLoop = false;
+	for (char abbreviation: input)
+		switch (abbreviation)
+		{
+		case '(':
+			assertThrow(!insideLoop, OptimizerException, "Nested parentheses not supported");
+			insideLoop = true;
+			break;
+		case ')':
+			assertThrow(insideLoop, OptimizerException, "Unbalanced parenthesis");
+			insideLoop = false;
+			break;
+		default:
+			assertThrow(
+				stepAbbreviationToNameMap().find(abbreviation) != stepAbbreviationToNameMap().end(),
+				OptimizerException,
+				"Invalid optimisation step abbreviation"
+			);
+		}
+	assertThrow(!insideLoop, OptimizerException, "Unbalanced parenthesis");
+
+	auto abbreviationsToSteps = [](string const& _sequence) -> vector<string>
+	{
+		vector<string> steps;
+		for (char abbreviation: _sequence)
+			steps.emplace_back(stepAbbreviationToNameMap().at(abbreviation));
+		return steps;
+	};
+
+	// The sequence has now been validated and must consist of pairs of segments that look like this: `aaa(bbb)`
+	// `aaa` or `(bbb)` can be empty. For example we consider a sequence like `fgo(aaf)Oo` to have
+	// four segments, the last of which is an empty parenthesis.
+	size_t currentPairStart = 0;
+	while (currentPairStart < input.size())
+	{
+		size_t openingParenthesis = input.find('(', currentPairStart);
+		size_t closingParenthesis = input.find(')', openingParenthesis);
+		size_t firstCharInside = (openingParenthesis == string::npos ? input.size() : openingParenthesis + 1);
+		solAssert((closingParenthesis == string::npos) == (closingParenthesis == string::npos), "");
```

Comment:
You probably want to change something in this assertion - weird that the compiler does not warn.
```suggestion
		solAssert((closingParenthesis == string::npos) == (closingParenthesis == string::npos), "");
```

Reply:
Right. My bad. It was supposed to ensure that there can be a closing parenthesis in front of you IFF there's a matching opening one too.

Fixed.